### PR TITLE
Fixes #525: Add eslint-plugin-promise and fix async/Promise issues

### LINKED ISSUE #525: Add eslint-plugin-promise and fix async/Promise issues
We have a bunch of improper uses of `async` (e.g., without `await`) and also a bunch of `Promise` bugs that will affect performance.

A first step on the path to fixing this would be to add some linting for Promises.  We can use https://github.com/xjamundx/eslint-plugin-promise, with the recommended rules defined.  I tested it locally, and it already finds 24 problems we can fix.

We should also eliminate all pure synchronous calls that have been wrapped with Promises unnecessarily.  In essence, if we don't need a Promise, let's get rid of it.

[This recent talk](https://www.youtube.com/watch?v=XV-u_Ow47s0&feature=emb_logo) is a good intro to some of the problems I'm referring to here.

### PR DESCRIPTION
**Pull request by c3ho**
## Type of Change

- **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

Added eslint rules regarding promises, we have promises that have catches and returns and some don't. These rules ensure consistency

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by c3ho** [comment]
Having issues with dealing with inactive-blog-filter.js. It returns a lot of promises.

File: .eslintrc.js [hunk 1]
```diff
@@ -23,6 +23,36 @@ module.exports = {
...
+     * usage of await or yield
+     * https://eslint.org/docs/rules/require-atomic-updates
+     */
+    'require-atomic-updates': ['error'],
+
+    /**
+     * Disallow async functions which have no await expression
+     * https://eslint.org/docs/rules/require-await
+     */
+    'require-awaut': ['error'],
```
Comment (manekenpix) [inline .eslintrc.js]:
```suggestion
    'require-await': ['error'],
```

Comment (manekenpix) [reply .eslintrc.js]: Also, this one seems to create a conflict with the rule in line 29 `no-return-await` in cases like [line 71 in storage.js](https://github.com/Seneca-CDOT/telescope/blob/cda82ab740464c04d50605bd6576382809381311/src/backend/utils/storage.js#L71).

Comment (c3ho) [reply .eslintrc.js]: I think we can probably fix line 71 with this

```
getPost: async guid => {
  const posts = await redis.getall(guid);
  return posts;
  },
```

**Comment by c3ho** [comment]
I think all the functions in storage.js are/should be async no?

**Comment by c3ho** [comment]
Running npm test will fail on the wiki-feed-parser.test
image

Otherwise it passes every test. Please go easy on me.

**Comment by manekenpix** [comment]
Tests pass for me.

File: .eslintrc.js [hunk 2]
```diff
@@ -34,5 +64,80 @@ module.exports = {
      * https://eslint.org/docs/rules/func-names
      */
     'linebreak-style': 'off',
+
+    /**
+     * The following are eslint rules from the promise-plugin
+     * https://github.com/xjamundx/eslint-plugin-promise
```
Comment (humphd) [inline .eslintrc.js]: Question: why not use the recommended set vs. manually enabling all these?

```
{
  "extends": ["plugin:promise/recommended"]
}
```

Are you using ones other than the recommended?

File: src/backend/index.js [hunk 3]
```diff
@@ -42,6 +42,7 @@ async function enqueueWikiFeed() {
 enqueueWikiFeed()
   .then(() => {
```
Comment (humphd) [inline src/backend/index.js]: Often this means you can just rely on the implicit return value you get with an arrow function.  I think this is better:

```
.then(() => feedWorker.start())
```

File: src/backend/utils/basic_analysis.js [hunk 4]
```diff
@@ -71,17 +71,17 @@ module.exports = function analyzeText(text_) {
...
-        res(analysis);
-      } else {
-        res(textInfo);
-      }
-      if (isValidString === undefined) {
-        rej(textInfo);
-      }
-    });
+  this.getAsyAnalysis = function() {
+    let result;
```
Comment (humphd) [inline src/backend/utils/basic_analysis.js]: Drop this `result` arg, and just use `return` on every line where you set `result` below.  [Return early, return often](https://blog.humphd.org/vocamus-1421/).

File: src/backend/utils/email-sender.js [hunk 5]
```diff
@@ -90,18 +90,16 @@ exports.sendMessage = async function(receipiants, subjectMessage, message) {
     );
     const allGood = this.verifyTransporter(transporter);
     if (!allGood) {
-      reject(new Error()); // Send promise.reject if an error occurs
+      logger.error(new Error());
```
Comment (humphd) [inline src/backend/utils/email-sender.js]: Get rid of the variable, add a message to the error so it's useful, throw.

```js
if(!this.verifyTransporter(transporter)) {
  throw new Error('email transport could not be verified');
}
```

File: src/backend/utils/github-url.js [hunk 6]
```diff
@@ -57,68 +57,66 @@ exports.getGithubUrlData = async incomingUrl => {
       ? `?access_token=${process.env.GITHUB_TOKEN}`
       : '';
 
-  return fetch(`${githubAPI}${subUrl}`)
-    .then(res => res.json())
-    .then(data => {
-      let fetchedData;
+  const fetchResult = await fetch(`${githubAPI}${subUrl}`);
```
Comment (humphd) [inline src/backend/utils/github-url.js]: `fetch` can throw, so can `.json()`, so we should wrap these in try/catch.

File: src/backend/utils/storage.js [hunk 7]
```diff
@@ -68,5 +68,8 @@ module.exports = {
 
   getPostsCount: () => redis.zcard(POSTS),
 
-  getPost: async guid => redis.hgetall(guid),
+  getPost: async guid => {
```
Comment (humphd) [inline src/backend/utils/storage.js]: No need for async/await here, this already returns a Promise.

```js
getPost: guid => redis.hgetall(guid)
```

File: test/email-sender.test.js [hunk 8]
```diff
@@ -12,7 +12,7 @@ test.skip('Tests if sendMessage resolves with expected info', async () => {
   const testRecipient = process.env.NODEMAILER_TESTRECIPIENT;
   const testReturnValue = emailsender.sendMessage(testRecipient, testSubjectMessage, testHTML);
   // Tests if the expected resolve is correct
-  expect(testReturnValue).resolves.toBe([testRecipient]);
+  await expect(testReturnValue).resolves.toBe([testRecipient]);
```
Comment (humphd) [inline test/email-sender.test.js]: Should this `await` not go with the call to `emailsender.sendMessage` on line 13 above?

File: test/spam-checker.test.js [hunk 9]
```diff
@@ -1,5 +1,6 @@
 const feedparser = require('feedparser-promised');
 const checkForSpam = require('../src/backend/utils/spam-checker');
+// const { logger } = require('../src/backend/utils/logger');
```
Comment (humphd) [inline test/spam-checker.test.js]: Remove this if not needed.

File: test/storage.test.js [hunk 10]
```diff
@@ -81,7 +82,13 @@ describe('Storage tests for posts', () => {
   };
 
   beforeAll(() => {
-    Promise.all([testPost, testPost2, testPost3].map(post => addPost(post)));
+    Promise.all([testPost, testPost2, testPost3].map(post => addPost(post)))
+      .then(() => {
```
Comment (humphd) [inline test/storage.test.js]: Why are we mixing async/await and .then?  Do we really need this .then at all?

File: .eslintrc.js (hunk 2 above)
Comment (humphd) [reply .eslintrc.js]: What if we do a bit of both here: Extend the recommended set, and for all those ones that aren't in recommended that you want from the [rules set](https://github.com/xjamundx/eslint-plugin-promise#rules), add them manually to the rules below.

File: src/backend/utils/basic_analysis.js [hunk 11]
```diff
@@ -71,18 +71,14 @@ module.exports = function analyzeText(text_) {
...
-        res(analysis);
-      } else {
-        res(textInfo);
-      }
-      if (isValidString === undefined) {
-        rej(textInfo);
-      }
-    });
+  this.getAsyAnalysis = function() {
+    if (isValidString === undefined) {
```
Comment (humphd) [inline src/backend/utils/basic_analysis.js]: It looks like this can all be simplified to:

```js
return isValidString ? analysis : textInfo;
```

Make sure I'm right (untested), but just reading this code, there's no reason for all these different code paths that I can see.

File: src/backend/utils/github-url.js [hunk 12]
```diff
@@ -56,69 +56,70 @@ exports.getGithubUrlData = async incomingUrl => {
     typeof process.env.GITHUB_TOKEN !== 'undefined'
       ? `?access_token=${process.env.GITHUB_TOKEN}`
       : '';
+  try {
+    const fetchResult = await fetch(`${githubAPI}${subUrl}`);
+    const data = await fetchResult.json();
```
Comment (humphd) [inline src/backend/utils/github-url.js]: Put your `catch` here vs. at the bottom, it's only these 2 lines that can throw.

File: src/backend/utils/github-url.js [hunk 13]
```diff
@@ -56,69 +56,70 @@ exports.getGithubUrlData = async incomingUrl => {
...
+        avatarURL,
+        body,
+        createdAt,
+        branch: ghUrl.branch,
+        repo: ghUrl.name,
+      };
+    }
+    return fetchedData;
+  } catch (error) {
+    throw new Error('Error fetching data');
```
Comment (humphd) [inline src/backend/utils/github-url.js]: If all you're going to do in a `.catch` is `throw`, there's no value in doing it, and you can just leave it off.  However, it might be nice to log the error here, before you throw it.  Also, I'd consider just re-throwing the actual error vs. making your own, and losing the original error message and stack.

Comment (humphd) [reply src/backend/utils/github-url.js]: What you do in  `src/backend/utils/inactive-blog-filter.js` below is what I mean.

File: src/backend/utils/storage.js [hunk 14]
```diff
@@ -68,5 +68,7 @@ module.exports = {
 
   getPostsCount: () => redis.zcard(POSTS),
 
-  getPost: async guid => redis.hgetall(guid),
+  getPost: guid => {
```
Comment (humphd) [inline src/backend/utils/storage.js]: You don't need the explicit `return` here:

```js
getPost: guid => redis.hgetall(guid);
```

A fat arrow function will automatically return if you give it an expression vs. a `{...}` body block.

File: src/backend/utils/word-counter.js [hunk 15]
```diff
@@ -1,5 +1,8 @@
+/**
+ * https://www.npmjs.com/package/wordcount
+ */
 const wordcount = require('wordcount');
 
-exports.wordCounter = async function wordCounter(blogText) {
-  return Promise.resolve(wordcount(blogText));
+exports.wordCounter = function wordCounter(blogText) {
```
Comment (humphd) [inline src/backend/utils/word-counter.js]: This file isn't doing anything.  We're pulling in the `wordcount` module, then creating a function called `wordCounter` that calls `wordcount`.  Let's just `git rm` this file, and change the caller(s) to use this module directly.

File: test/inactive-blog-filter.test.js [hunk 16]
```diff
@@ -36,31 +36,20 @@ describe('Redlisted feed checking', () => {
...
-          expect(isRedlisted).toBe(false);
-        })
-        .catch(err => {
-          log.error(err.message);
-          throw err;
-        });
-    });
+    await Promise.all(
+      badURLs.map(async url => {
+        // const inactiveFilterCheck = await inactiveFilter.check(url);
```
Comment (humphd) [inline test/inactive-blog-filter.test.js]: Why is this line here?  Remove or uncomment, please.

File: test/inactive-blog-filter.test.js [hunk 17]
```diff
@@ -129,54 +118,57 @@ describe('Testing of update()', () => {
...
+      const data = await inactiveFilter.update(
+        pathToInactiveFeeds,
+        pathToMockRedList,
+        mockFeedParser
+      );
+      expect(data.length).toEqual(inactiveRedlist.length);
+      for (let i = 0; i < inactiveRedlist.length; i += 1) {
+        expect(data[i].url).toBe(inactiveRedlist[i].url);
+      }
+      return data;
```
Comment (humphd) [inline test/inactive-blog-filter.test.js]: Who are we returning `data` to here?  Can we get rid of this?

File: test/inactive-blog-filter.test.js [hunk 18]
```diff
@@ -129,54 +118,57 @@ describe('Testing of update()', () => {
...
+    } catch (err) {
+      deleteMockRedlist();
+      log.error(err.message);
```
Comment (humphd) [inline test/inactive-blog-filter.test.js]: I think in unit tests, it's better to just let things throw, and not bother with catching, unless we want to test something in the error handling path.  The test harness will blow up if there's an exception, and print it, so we don't have to.

File: test/posts.test.js [hunk 19]
```diff
@@ -24,8 +25,12 @@ describe('test /posts endpoint', () => {
     };
   });
 
-  beforeAll(() => {
-    Promise.all(posts.map(post => addPost(post)));
+  beforeAll(async () => {
+    try {
```
Comment (humphd) [inline test/posts.test.js]: Don't bother with try/catch here.  Let it fail loudly in tests.

File: test/storage.test.js [hunk 20]
```diff
@@ -80,8 +81,13 @@ describe('Storage tests for posts', () => {
     site: 'foo',
   };
 
-  beforeAll(() => {
-    Promise.all([testPost, testPost2, testPost3].map(post => addPost(post)));
+  beforeAll(async () => {
+    try {
```
Comment (humphd) [inline test/storage.test.js]: Drop this try/catch.

File: src/backend/utils/github-url.js [hunk 21]
```diff
@@ -1,7 +1,9 @@
 require('../lib/config');
 const parseGithubUrl = require('parse-github-url');
 const fetch = require('node-fetch');
+const { logger } = require('./logger');
 
+const log = logger.child({ module: 'github-url' });
```
Comment (humphd) [inline src/backend/utils/github-url.js]: Let's not do this.  The docs suggest it, but for our needs, it's not really necessary to isolate logs into sub-modules at this stage.  Just use `logger.info()`, etc.

File: src/backend/utils/inactive-blog-filter.js [hunk 22]
```diff
@@ -38,17 +38,14 @@ function dateDiff(postDate) {
...
-      return Promise.resolve(redList.some(isRedlisted.bind(null, feedUrl)));
-    })
-    .catch(error => {
-      log.error({ error }, `failed to read ${redlistPath}`);
-      return Promise.reject(error);
-    });
+  try {
+    const redListRaw = await fsPromises.readFile(redlistPath, 'utf-8');
+    const redList = [].concat(JSON.parse(redListRaw));
+    return redList.some(isRedlisted.bind(null, feedUrl));
```
Comment (humphd) [inline src/backend/utils/inactive-blog-filter.js]: Could this not be `return redList.some(redItem => isRedlisted(feedUrl, redItem));`?  It doesn't seem like we need to `.bind()` here.

File: .eslintrc.js (hunk 2 above)
Comment (humphd) [reply .eslintrc.js]: This looks undone to me, are you going to modify this further?

**Comment by humphd** [comment]
@c3ho did you ever fix your error above in https://github.com/Seneca-CDOT/telescope/pull/527#issuecomment-573290446?  It means you need to rebuild some binary dependency that was linked against header files for a different version of node/v8.  If you update node, you sometimes have to rebuild.  There's a command to do it, or you can `rm -fr node_modules && npm install`.

**Comment by c3ho** [comment]
@humphd it is flaky for me on Windows. Sometimes it passes with no issues, sometimes I run into the error. I've done what you've usually suggested in OSD600 removing the dependency folders and everything. I've even tried installing nvm on my windows machine. I tested this on Mac and there were no issues.

File: test/posts.test.js [hunk 23]
```diff
@@ -24,8 +24,8 @@ describe('test /posts endpoint', () => {
     };
   });
 
-  beforeAll(() => {
-    Promise.all(posts.map(post => addPost(post)));
+  beforeAll(async () => {
```
Comment (humphd) [inline test/posts.test.js]: Jest's functions work async if you return a Promise, so you can change this to:

```js
beforeAll(() => Promise.all(posts.map(post => addPost(post))));
```

```

File: test/posts.test.js [hunk 24]
```diff
@@ -120,8 +120,8 @@ describe('test /posts/:guid responses', () => {
 
   // add the post to the storage
   beforeAll(async () => {
```
Comment (humphd) [inline test/posts.test.js]:
```js
beforeAll(() => Promise.all([addedPost1.save(), addedPost2.save()]));
```

File: test/storage.test.js [hunk 25]
```diff
@@ -80,8 +80,8 @@ describe('Storage tests for posts', () => {
     site: 'foo',
   };
 
-  beforeAll(() => {
-    Promise.all([testPost, testPost2, testPost3].map(post => addPost(post)));
+  beforeAll(async () => {
```
Comment (humphd) [inline test/storage.test.js]:
```js
beforeAll(() => Promise.all([testPost, testPost2, testPost3].map(post => addPost(post))));
```

File: test/posts.test.js (hunk 24 above)
Comment (humphd) [reply test/posts.test.js]: This one got missed.

**Comment by c3ho** [comment]
@humphd if this is good to go, I'm going to rebase again and resubmit, VSC was going weird on me and added your recent code as part of my PR

**Comment by c3ho** [comment]
@humphd its failing on this one 
image

Is the fix just a simple ```await``` in front of ```queue.add()``` (line 33)?

**Comment by humphd** [comment]
I filed https://github.com/Seneca-CDOT/telescope/issues/567 to deal with the problem of knowing whether something returns a `Promise` or not.

**Comment by humphd** [comment]
Yeah, merged!
